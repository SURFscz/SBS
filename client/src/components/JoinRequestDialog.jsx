import React from "react";

import I18n from "i18n-js";
import "./JoinRequestDialog.scss";
import Button from "./Button";
import InputField from "./InputField";
import {isEmpty} from "../utils/Utils";
import {joinRequestForCollaboration} from "../api";
import {ReactComponent as InformationIcon} from "../icons/informational.svg";
import DOMPurify from "dompurify";
import {AlertType, Modal} from "@surfnet/sds";

export default class JoinRequestDialog extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            motivation: "",
            submitted: false,
        }
    }

    submit = () => {
        const {collaboration,} = this.props;
        joinRequestForCollaboration({...this.state, collaborationId: collaboration.id})
            .then(() => this.setState({submitted: true}));
    }

    gotoHome = () => {
        const {refresh} = this.props;
        refresh(() => setTimeout(() => this.props.history.push("/home/joinrequests"), 75));
    }

    renderForm = (collaboration, motivation) => {
        return (
            <div>
                <InputField name={I18n.t("registration.motivation", {name: collaboration.name})}
                            value={motivation}
                            multiline={true}
                            placeholder={I18n.t("registration.motivationPlaceholder")}
                            onChange={e => this.setState({motivation: e.target.value})}/>
            </div>
        );
    }

    renderFeedback = collaboration => {
        return (
            <div>
                <section className="explanation informational">
                    <InformationIcon/>
                    <span dangerouslySetInnerHTML={{
                        __html: DOMPurify.sanitize(I18n.t("registration.feedback.info", {
                            name: collaboration.name
                        }))
                    }}/>
                </section>
                <section className="actions">
                    <Button txt={I18n.t("confirmationDialog.ok")}
                            onClick={this.gotoHome}/>
                </section>
            </div>
        )
            ;
    }

    content = (submitted, collaboration, motivation) => {
        return (
            <div className="join-request-form">
                {!submitted &&
                this.renderForm(collaboration, motivation, close)}
            </div>
        );
    }

    render() {
        const {collaboration, isOpen = false, close} = this.props;
        const {motivation, submitted} = this.state;
        if (!isOpen) {
            return null;
        }
        const subTitle = submitted ? I18n.t("registration.feedback.info", {name: collaboration.name}) :
            I18n.t("registration.explanation", {name: collaboration.name});
        return (
            <Modal
                confirm={submitted ? this.gotoHome : this.submit}
                cancel={submitted ? null : close}
                alertType={AlertType.Info}
                subTitle={subTitle}
                children={this.content(submitted, collaboration, motivation)}
                title={I18n.t("registration.title", {name: collaboration.name})}
                cancelButtonLabel={submitted ? null : I18n.t("forms.cancel")}
                confirmationButtonLabel={submitted ? I18n.t("confirmationDialog.ok") : I18n.t("forms.request")}
                confirmDisabled={isEmpty(motivation)}
                question={null}/>
        );
    }

}

