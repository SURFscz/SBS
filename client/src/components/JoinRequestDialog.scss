@import "../stylesheets/vars.scss";


.join-request-dialog-overlay {
    position: fixed;
    display: flex;
    align-items: center;
    justify-content: center;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(255, 255, 255, 0.85);
    z-index: 5;

    .join-request-dialog-content {
        padding: 40px 60px;

        @media (max-width: $compact-medium) {
            padding: 20px;
        }

        background-color: white;
        overflow: auto;
        border-radius: $br;
        outline: none;
        z-index: 5;
        display: flex;
        flex-direction: column;
        box-shadow: 1px 2px 2px 0 rgba(0, 0, 0, 0.3);
        border: 1px solid $lighter-grey;


        h1 {
            margin-bottom: 30px;
        }

        .join-request-disclaimer {
            p {
                margin-bottom: 20px
            }
        }

        section.explanation {
            background-color: $yellow;
            padding: 15px;
            border-radius: $br;

            &.informational {
                display: flex;
                align-items: center;
                background-color: #95d7e4;

                svg {
                    margin-right: 15px;
                    width: 64px;
                    height: auto;
                }
            }
        }

        textarea, .checkbox {
            margin-bottom: 25px;
        }

    }
}

.ReactModalPortal .join-request-dialog-overlay {
    transition: opacity 250ms ease-in-out;

    &--before-close {
        opacity: 0;
    }
}

.ReactModalPortal .join-request-dialog-content {
    transition: top 250ms ease-in-out, bottom 250ms ease-in-out;
    width: 860px;
    max-height: 85vh;

    @media (max-width: $compact-medium) {
        width: 280px;
    }

    &--before-close {
        top: -50%;
        bottom: 100%;
    }
}
