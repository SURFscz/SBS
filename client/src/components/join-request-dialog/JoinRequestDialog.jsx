import React from "react";

import I18n from "../../locale/I18n";
import "./JoinRequestDialog.scss";
import InputField from "../input-field/InputField";
import {isEmpty} from "../../utils/Utils";
import {joinRequestForCollaboration} from "../../api";
import {Modal} from "@surfnet/sds";
import RadioButtonGroup from "../_redesign/RadioButtonGroup";
import DOMPurify from "dompurify";
import CollaborationAupAcceptance from "../collaboration-aup-acceptance/CollaborationAupAcceptance";
import OrganisationAupAcceptance from "../OrganisationAupAcceptance";
import {aupData} from "../../utils/Aups";

export default class JoinRequestDialog extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            motivation: "",
            reason: "invited",
            submitted: false,
            organisationDisabled: false,
            disabled: false,
            services: []
        }
    }

    componentDidMount = () => {
        const {collaboration, user} = this.props;
        const {
            services,
            hasServices,
            requiresOrganisationAup,
            allServiceAupsAgreedOn
        } = aupData(user, collaboration);

        this.setState({
            disabled: hasServices && !allServiceAupsAgreedOn,
            organisationDisabled: requiresOrganisationAup,
            requiresOrganisationAup: requiresOrganisationAup,
            services: services,
            allServiceAupsAgreedOn: allServiceAupsAgreedOn,

        });
    }

    submit = () => {
        const {collaboration} = this.props;
        const {reason, motivation} = this.state;
        const motivationValue = reason === "other" ? motivation : I18n.t(`joinRequest.${reason}`);
        joinRequestForCollaboration({motivation: motivationValue, collaborationId: collaboration.id})
            .then(() => this.setState({submitted: true}));
    }

    gotoHome = () => {
        const {refresh} = this.props;
        refresh(() => setTimeout(() => this.props.history.push("/home/joinrequests"), 75));
    }

    aupAcceptance = () => {
        const {adminEmails, serviceEmails, collaboration} = this.props;
        const {requiresOrganisationAup, disabled, organisationDisabled, services, allServiceAupsAgreedOn} = this.state;
        return (
            <>
                <CollaborationAupAcceptance services={services}
                                            disabled={disabled}
                                            serviceEmails={serviceEmails}
                                            setDisabled={() => this.setState({disabled: !this.state.disabled})}
                                            allServiceAupsAgreedOn={allServiceAupsAgreedOn}
                                            children={services.length > 0 ?
                                                <h4 className="aup-services">{I18n.t("models.collaboration.joinRequestServices", {
                                                    nbr: services.length,
                                                    name: collaboration.name
                                                })}</h4> :
                                                <h4 className="aup-services">{I18n.t("models.collaboration.noServicesYet")}</h4>}
                />
                {requiresOrganisationAup &&
                    <OrganisationAupAcceptance adminEmails={adminEmails}
                                               disabled={organisationDisabled}
                                               setDisabled={() => this.setState({organisationDisabled: !this.state.organisationDisabled})}
                                               organisation={collaboration.organisation}/>
                }
            </>
        );
    }

    content = (submitted, collaboration, motivation, reason) => {
        return (
            <>
                <div className="join-request-form">
                    {!submitted &&
                        <div>
                            <RadioButtonGroup name={"reason"}
                                              label={I18n.t("joinRequest.why")}
                                              value={reason}
                                              horizontal={false}
                                              values={["invited", "projectMember", "other"]}
                                              onChange={value => this.setState({reason: value})}
                                              labelResolver={label => I18n.t(`joinRequest.${label}`)}/>
                            {reason === "other" &&
                                <InputField name={I18n.t("registration.motivation", {name: collaboration.name})}
                                            value={motivation}
                                            multiline={true}
                                            placeholder={I18n.t("registration.motivationPlaceholder")}
                                            onChange={e => this.setState({motivation: e.target.value})}/>}
                        </div>}
                    {submitted && <div>
                    <span
                        dangerouslySetInnerHTML={{__html: DOMPurify.sanitize(I18n.t("registration.feedback.info", {name: collaboration.name}))}}/>
                    </div>}

                </div>
                {!submitted && this.aupAcceptance()}
            </>
        );
    }

    render() {
        const {collaboration, isOpen = false, close} = this.props;
        const {motivation, submitted, reason, disabled, organisationDisabled} = this.state;
        if (!isOpen) {
            return null;
        }

        return (
            <Modal
                confirm={submitted ? this.gotoHome : this.submit}
                cancel={submitted ? null : close}
                className="join-request-dialog"
                children={this.content(submitted, collaboration, motivation, reason)}
                title={I18n.t("registration.title", {name: collaboration.name})}
                cancelButtonLabel={submitted ? null : I18n.t("forms.cancel")}
                confirmationButtonLabel={submitted ? I18n.t("confirmationDialog.ok") : I18n.t("forms.request")}
                confirmDisabled={disabled || organisationDisabled || (isEmpty(motivation) && reason === "other")}
                question={null}/>
        );
    }

}
