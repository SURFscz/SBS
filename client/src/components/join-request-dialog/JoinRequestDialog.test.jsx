import React from "react";
import {render} from "@testing-library/react";
import JoinRequestDialog from "./JoinRequestDialog";

jest.mock("../../locale/I18n", () => ({
    t: (key, opts) => key + (opts ? JSON.stringify(opts) : ""),
}));

jest.mock("../../api", () => ({
    joinRequestForCollaboration: jest.fn(() => Promise.resolve()),
}));

jest.mock("../../utils/Aups", () => ({
    aupData: () => ({
        services: [],
        hasServices: false,
        requiresOrganisationAup: false,
        allServiceAupsAgreedOn: true,
    }),
}));

describe("JoinRequestDialog", () => {
    it("matches snapshot when open", () => {
        const props = {
            isOpen: true,
            close: jest.fn(),
            refresh: jest.fn(),
            history: {push: jest.fn()},
            collaboration: {id: 1, name: "TestCollab", organisation: {}},
            user: {},
            adminEmails: [],
            serviceEmails: [],
        };
        const {asFragment} = render(<JoinRequestDialog {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
