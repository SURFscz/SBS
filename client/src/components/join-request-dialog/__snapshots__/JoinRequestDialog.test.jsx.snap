// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinRequestDialog matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="sds--modal sds--backdrop"
  >
    <div
      class="sds--modal--container  join-request-dialog"
    >
      <div
        class="sds--modal--title "
      >
        <p
          class="sds--text--h4"
        >
          registration.title{"name":"TestCollab"}
        </p>
        <span
          class="sds--modal--close"
        >
          <svg
            height="13.5"
            viewBox="0 0 13.5 13.5"
            width="13.5"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m.8 12.8 12-12m0 12L.8.8"
              fill="none"
              stroke="#000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
        </span>
      </div>
      <div
        class="sds--modal--content"
      >
        <div
          class="sds--text--rich"
        >
          <div
            class="join-request-form"
          >
            <div>
              <div
                class="sds--radio-options"
              >
                <label
                  class="sds--tooltip-parent sds--radio-options-label"
                  for="reason"
                >
                  <span>
                    joinRequest.why
                  </span>
                </label>
                <div
                  class="sds--text-field-container column"
                >
                  <div
                    class="option-0"
                  >
                    <input
                      checked=""
                      id="reason_invited"
                      name="reason"
                      type="radio"
                    />
                    <label
                      class="sds--radio-options-label"
                      for="reason_invited"
                    >
                      <span
                        class="sds--radio--visual"
                      />
                      <span
                        class="sds--radio--text"
                      >
                        joinRequest.invited
                      </span>
                    </label>
                  </div>
                  <div
                    class="option-1"
                  >
                    <input
                      id="reason_projectMember"
                      name="reason"
                      type="radio"
                    />
                    <label
                      class="sds--radio-options-label"
                      for="reason_projectMember"
                    >
                      <span
                        class="sds--radio--visual"
                      />
                      <span
                        class="sds--radio--text"
                      >
                        joinRequest.projectMember
                      </span>
                    </label>
                  </div>
                  <div
                    class="option-2"
                  >
                    <input
                      id="reason_other"
                      name="reason"
                      type="radio"
                    />
                    <label
                      class="sds--radio-options-label"
                      for="reason_other"
                    >
                      <span
                        class="sds--radio--visual"
                      />
                      <span
                        class="sds--radio--text"
                      >
                        joinRequest.other
                      </span>
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mod-aup-services"
          >
            <h4
              class="aup-services"
            >
              models.collaboration.noServicesYet
            </h4>
            <div
              class="services"
            >
              <span>
                welcomeDialog.noServices
              </span>
            </div>
          </div>
        </div>
        <div
          class="sds--modal--actions sds--actions--outer sds--space--top--5"
        >
          <button
            class="sds--btn sds--btn--secondary     "
            type="button"
          >
            <span
              class="textual"
            >
              forms.cancel
            </span>
          </button>
          <button
            class="sds--btn sds--btn--primary     "
            type="button"
          >
            <span
              class="textual"
            >
              forms.request
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
